# Feature Engineering 이란
도메인 지식과 창의성을 바탕으로 데이터셋에 존재하는 Feature들을 재조합하여 새로운 Feature를 만드는 것

간단한 연산일 수도 있고 복잡한 계산이 필요할 수도 있음

분석 모델의 더 좋은 퍼포먼스를 위하여 더 새롭고 더 의미있는 패턴을 제공하는 것이 궁극적인 목적

In [2]:
import pandas as pd
df = pd.read_csv('https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/kt%26g/kt%26g.csv')

df

분기     매출액   영업이익 영업이익(발표기준) 세전계속사업이익  당기순이익 당기순이익(지배)  당기순이익(비지배)  \
0   19/6  12,578  3,991      3,991    4,452  3,245     3,228          17   
1   19/9  13,222  3,825      3,825    4,503  3,221     3,208          13   
2  19/12  11,982  2,523      2,523    1,794  1,183     1,198         -15   
3   20/3  11,784  3,150      3,150    4,156  2,939     2,930           9   
4   20/6  13,188  3,947      3,947    4,020  2,935     2,931           4   

      자산총계    부채총계  ... 투자활동현금흐름 재무활동현금흐름  영업이익률   순이익률 ROE(%) ROA(%)   부채비율  \
0  108,464  25,940  ...   -1,878   -4,919  31.73  25.80  12.35   9.37  31.43   
1  106,314  20,691  ...      -83     -329  28.93  24.36  12.67   9.95  24.17   
2  107,121  20,062  ...     -501     -129  21.06   9.87  12.39   9.96  23.04   
3  108,594  24,862  ...   -1,392      -86  26.73  24.94  13.05   9.79  29.69   
4  110,282  23,386  ...     -214   -5,501  29.93  22.25  12.20   9.40  26.91   

    자본유보율  EPS(원)  PER(배)  
0  803.22   2,351   13.77  
1  836.86   2,337   13.87  
2  850.82     873   12.42  
3  823.16   2,134    9.71  
4  837.40   2,135   10.46  

[5 rows x 25 columns]

In [3]:
df.dtypes

분기             object
매출액            object
영업이익           object
영업이익(발표기준)     object
세전계속사업이익       object
당기순이익          object
당기순이익(지배)      object
당기순이익(비지배)      int64
자산총계           object
부채총계           object
자본총계           object
자본총계(지배)       object
자본총계(비지배)     float64
자본금            object
영업활동현금흐름       object
투자활동현금흐름       object
재무활동현금흐름       object
영업이익률         float64
순이익률          float64
ROE(%)        float64
ROA(%)        float64
부채비율          float64
자본유보율         float64
EPS(원)         object
PER(배)        float64
dtype: object

In [4]:
df['자본총계(비지배)']

0      NaN
1      NaN
2      NaN
3    562.0
4    566.0
Name: 자본총계(비지배), dtype: float64

## Na, Null, NaN, 0, Undefined 의 차이는?

<br>

## Feature Engineering

In [5]:
# J-value = ROE + ROA
df['j-value'] = df['ROE(%)'] + df['ROA(%)']
df

분기     매출액   영업이익 영업이익(발표기준) 세전계속사업이익  당기순이익 당기순이익(지배)  당기순이익(비지배)  \
0   19/6  12,578  3,991      3,991    4,452  3,245     3,228          17   
1   19/9  13,222  3,825      3,825    4,503  3,221     3,208          13   
2  19/12  11,982  2,523      2,523    1,794  1,183     1,198         -15   
3   20/3  11,784  3,150      3,150    4,156  2,939     2,930           9   
4   20/6  13,188  3,947      3,947    4,020  2,935     2,931           4   

      자산총계    부채총계  ... 재무활동현금흐름  영업이익률   순이익률 ROE(%) ROA(%)   부채비율   자본유보율  \
0  108,464  25,940  ...   -4,919  31.73  25.80  12.35   9.37  31.43  803.22   
1  106,314  20,691  ...     -329  28.93  24.36  12.67   9.95  24.17  836.86   
2  107,121  20,062  ...     -129  21.06   9.87  12.39   9.96  23.04  850.82   
3  108,594  24,862  ...      -86  26.73  24.94  13.05   9.79  29.69  823.16   
4  110,282  23,386  ...   -5,501  29.93  22.25  12.20   9.40  26.91  837.40   

   EPS(원)  PER(배)  j-value  
0   2,351   13.77    21.72  
1   2,337   13.87    22.62  
2     873   12.42    22.35  
3   2,134    9.71    22.84  
4   2,135   10.46    21.60  

[5 rows x 26 columns]

In [6]:
df['자산'] = df['부채총계'] + df['자본총계']
df

분기     매출액   영업이익 영업이익(발표기준) 세전계속사업이익  당기순이익 당기순이익(지배)  당기순이익(비지배)  \
0   19/6  12,578  3,991      3,991    4,452  3,245     3,228          17   
1   19/9  13,222  3,825      3,825    4,503  3,221     3,208          13   
2  19/12  11,982  2,523      2,523    1,794  1,183     1,198         -15   
3   20/3  11,784  3,150      3,150    4,156  2,939     2,930           9   
4   20/6  13,188  3,947      3,947    4,020  2,935     2,931           4   

      자산총계    부채총계  ...  영업이익률   순이익률  ROE(%) ROA(%)   부채비율   자본유보율 EPS(원)  \
0  108,464  25,940  ...  31.73  25.80   12.35   9.37  31.43  803.22  2,351   
1  106,314  20,691  ...  28.93  24.36   12.67   9.95  24.17  836.86  2,337   
2  107,121  20,062  ...  21.06   9.87   12.39   9.96  23.04  850.82    873   
3  108,594  24,862  ...  26.73  24.94   13.05   9.79  29.69  823.16  2,134   
4  110,282  23,386  ...  29.93  22.25   12.20   9.40  26.91  837.40  2,135   

   PER(배)  j-value            자산  
0   13.77    21.72  25,94082,524  
1   13.87    22.62  20,69185,623  
2   12.42    22.35  20,06287,059  
3    9.71    22.84  24,86283,732  
4   10.46    21.60  23,38686,896  

[5 rows x 27 columns]

# String 연산 문제 해결

1. 문자를 숫자로 바꾸기 위해 숫자가 아닌 부분을 제거
2. 문자를 숫자로 형변환

## string replace

replace의 사용법

**string variable**.replace(“삭제할 글자“, ‘’) 의 형태로 사용 ( 공백으로 대치 )

s.replace(',', '')

In [7]:
testString = '25,970'
testString.replace(',','')

'25970'

In [8]:
testString

'25,970'

## Type cating and as Function

In [9]:
testString = testString.replace(',','')
int(testString)

25970

In [10]:
# 입력된 문자열에 대해서 같은 작업을 하는 함수 작성
def toInt(string):
    return int(string.replace(',',''))

In [11]:
# 예시 데이터를 바탕으로 함수를 테스트
toInt('25,970')

25970

In [12]:
type(toInt('25,970'))

int

# Apply

데이터의 모든 문자열에 대해서 일일히 toInt 함수를 반복할 수는 없음
대신 column 단위로 수행

## apply 사용법
1. apply 안에 들어갈 함수를 선언
2. column에 apply 적용.

In [13]:
df['자산2'] = df['자산'].apply(toInt)
df

분기     매출액   영업이익 영업이익(발표기준) 세전계속사업이익  당기순이익 당기순이익(지배)  당기순이익(비지배)  \
0   19/6  12,578  3,991      3,991    4,452  3,245     3,228          17   
1   19/9  13,222  3,825      3,825    4,503  3,221     3,208          13   
2  19/12  11,982  2,523      2,523    1,794  1,183     1,198         -15   
3   20/3  11,784  3,150      3,150    4,156  2,939     2,930           9   
4   20/6  13,188  3,947      3,947    4,020  2,935     2,931           4   

      자산총계    부채총계  ...   순이익률 ROE(%)  ROA(%)   부채비율   자본유보율 EPS(원) PER(배)  \
0  108,464  25,940  ...  25.80  12.35    9.37  31.43  803.22  2,351  13.77   
1  106,314  20,691  ...  24.36  12.67    9.95  24.17  836.86  2,337  13.87   
2  107,121  20,062  ...   9.87  12.39    9.96  23.04  850.82    873  12.42   
3  108,594  24,862  ...  24.94  13.05    9.79  29.69  823.16  2,134   9.71   
4  110,282  23,386  ...  22.25  12.20    9.40  26.91  837.40  2,135  10.46   

   j-value            자산         자산2  
0    21.72  25,94082,524  2594082524  
1    22.62  20,69185,623  2069185623  
2    22.35  20,06287,059  2006287059  
3    22.84  24,86283,732  2486283732  
4    21.60  23,38686,896  2338686896  

[5 rows x 28 columns]

In [14]:
df['부채총계'] = df['부채총계'].apply(toInt)
df['자본총계'] = df['자본총계'].apply(toInt)

df['자산'] = df['부채총계'] + df['자본총계']

df

분기     매출액   영업이익 영업이익(발표기준) 세전계속사업이익  당기순이익 당기순이익(지배)  당기순이익(비지배)  \
0   19/6  12,578  3,991      3,991    4,452  3,245     3,228          17   
1   19/9  13,222  3,825      3,825    4,503  3,221     3,208          13   
2  19/12  11,982  2,523      2,523    1,794  1,183     1,198         -15   
3   20/3  11,784  3,150      3,150    4,156  2,939     2,930           9   
4   20/6  13,188  3,947      3,947    4,020  2,935     2,931           4   

      자산총계   부채총계  ...   순이익률 ROE(%)  ROA(%)   부채비율   자본유보율 EPS(원) PER(배)  \
0  108,464  25940  ...  25.80  12.35    9.37  31.43  803.22  2,351  13.77   
1  106,314  20691  ...  24.36  12.67    9.95  24.17  836.86  2,337  13.87   
2  107,121  20062  ...   9.87  12.39    9.96  23.04  850.82    873  12.42   
3  108,594  24862  ...  24.94  13.05    9.79  29.69  823.16  2,134   9.71   
4  110,282  23386  ...  22.25  12.20    9.40  26.91  837.40  2,135  10.46   

   j-value      자산         자산2  
0    21.72  108464  2594082524  
1    22.62  106314  2069185623  
2    22.35  107121  2006287059  
3    22.84  108594  2486283732  
4    21.60  110282  2338686896  

[5 rows x 28 columns]

# Exercise

# 1.Feature Engineering

* 각 분기별 영업이익률을 직접 계산

* Feature 이름은 영업이익률2

* 아래 데이터를 참조하여 18/19년도에 대해서 Price to Dream Ratio (PDR)을 계산

* 변수 이름은 df2, feature 이름은 PDR로, df2['PDR']을 통해 ~ 값이 확인 되어야 함. 아래의 전제조건을 활용하라.

* 점유율 (일반담배 기준)
 - 18년 : 62%
 - 19년 : 63.5%
* 시장의 규모
 - 18년 : 15조
 - 19년 : 15조

In [15]:
url = 'https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/kt%26g/kt%26g_0.csv'
headers = ['분기', '매출액', '영업이익', '영업이익(발표기준)', '세전계속사업이익',
           '당기순이익', '당기순이익(지배)', '당기순이익(비지배)', '자산총계', '부채총계',
           '자본총계', '자본총계(지배)', '자본총계(비지배)', '자본금', '영업활동현금흐름',
           '투자활동현금흐름', '재무활동현금흐름', '영업이익률', '순이익률', 'ROE(%)',
           'ROA(%)', '부채비율', '자본유보율', 'EPS(원)', 'PER(배)']

In [16]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings(action='ignore')

df = pd.read_csv(url, names=headers, header=None)

df

분기     매출액   영업이익 영업이익(발표기준) 세전계속사업이익  당기순이익 당기순이익(지배)  당기순이익(비지배)  \
0   19/6  12,578  3,991      3,991    4,452  3,245     3,228          17   
1   19/9  13,222  3,825      3,825    4,503  3,221     3,208          13   
2  19/12  11,982  2,523      2,523    1,794  1,183     1,198         -15   
3   20/3  11,784  3,150      3,150    4,156  2,939     2,930           9   
4   20/6  13,188  3,947      3,947    4,020  2,935     2,931           4   

      자산총계    부채총계  ... 투자활동현금흐름 재무활동현금흐름  영업이익률   순이익률 ROE(%) ROA(%)   부채비율  \
0  108,464  25,940  ...   -1,878   -4,919  31.73  25.80  12.35   9.37  31.43   
1  106,314  20,691  ...      -83     -329  28.93  24.36  12.67   9.95  24.17   
2  107,121  20,062  ...     -501     -129  21.06   9.87  12.39   9.96  23.04   
3  108,594  24,862  ...   -1,392      -86  26.73  24.94  13.05   9.79  29.69   
4  110,282  23,386  ...     -214   -5,501  29.93  22.25  12.20   9.40  26.91   

    자본유보율  EPS(원)  PER(배)  
0  803.22   2,351   13.77  
1  836.86   2,337   13.87  
2  850.82     873   12.42  
3  823.16   2,134    9.71  
4  837.40   2,135   10.46  

[5 rows x 25 columns]

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 25 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   분기          5 non-null      object 
 1   매출액         5 non-null      object 
 2   영업이익        5 non-null      object 
 3   영업이익(발표기준)  5 non-null      object 
 4   세전계속사업이익    5 non-null      object 
 5   당기순이익       5 non-null      object 
 6   당기순이익(지배)   5 non-null      object 
 7   당기순이익(비지배)  5 non-null      int64  
 8   자산총계        5 non-null      object 
 9   부채총계        5 non-null      object 
 10  자본총계        5 non-null      object 
 11  자본총계(지배)    5 non-null      object 
 12  자본총계(비지배)   2 non-null      float64
 13  자본금         5 non-null      object 
 14  영업활동현금흐름    5 non-null      object 
 15  투자활동현금흐름    5 non-null      object 
 16  재무활동현금흐름    5 non-null      object 
 17  영업이익률       5 non-null      float64
 18  순이익률        5 non-null      float64
 19  ROE(%)      5 non-null      float

데이터의 ' , ' 때문에 일반적인 형변환이 불가, ',' 를 제거해 줘야 함

영업이익률 = 영업이익 / 매출액 * 100

In [18]:
df["영업이익"] = pd.to_numeric(df["영업이익"].str.replace(',',''))

df['매출액'] = pd.to_numeric(df["매출액"].str.replace(',',''))

df['영업이익률2'] = df['영업이익'] / df['매출액'] * 100

df

분기    매출액  영업이익 영업이익(발표기준) 세전계속사업이익  당기순이익 당기순이익(지배)  당기순이익(비지배)  \
0   19/6  12578  3991      3,991    4,452  3,245     3,228          17   
1   19/9  13222  3825      3,825    4,503  3,221     3,208          13   
2  19/12  11982  2523      2,523    1,794  1,183     1,198         -15   
3   20/3  11784  3150      3,150    4,156  2,939     2,930           9   
4   20/6  13188  3947      3,947    4,020  2,935     2,931           4   

      자산총계    부채총계  ... 재무활동현금흐름  영업이익률   순이익률 ROE(%) ROA(%)   부채비율   자본유보율  \
0  108,464  25,940  ...   -4,919  31.73  25.80  12.35   9.37  31.43  803.22   
1  106,314  20,691  ...     -329  28.93  24.36  12.67   9.95  24.17  836.86   
2  107,121  20,062  ...     -129  21.06   9.87  12.39   9.96  23.04  850.82   
3  108,594  24,862  ...      -86  26.73  24.94  13.05   9.79  29.69  823.16   
4  110,282  23,386  ...   -5,501  29.93  22.25  12.20   9.40  26.91  837.40   

   EPS(원)  PER(배)     영업이익률2  
0   2,351   13.77  31.730005  
1   2,337   13.87  28.929058  
2     873   12.42  21.056585  
3   2,134    9.71  26.731161  
4   2,135   10.46  29.928723  

[5 rows x 26 columns]

PDR = 시가총액 / (시장 규모 * 시장점유율)

구분 종가 발행주식수 시가총액 시장점유율 PDR 0 18년 101500 137292497 13935188445500 62.0 ~ 1 19년 93800 137292497 12878036218600 63.5 ~

In [19]:
df2 = pd.DataFrame({'구분' : ['18년', '19년'],
                    '종가' : [101500, 93800 ],
                    '발행주식수' : [137292497	,137292497] ,
                    '시가총액' : [13935188445500, 12878036218600	],
                    '시장점유율' : [62.0, 63.5]
                    })
df2

구분      종가      발행주식수            시가총액  시장점유율
0  18년  101500  137292497  13935188445500   62.0
1  19년   93800  137292497  12878036218600   63.5

In [20]:
df2['PDR'] = df2['시가총액'] / (15000000000000 * df2['시장점유율']) * 100
df2['PDR']

0    1.498407
1    1.352025
Name: PDR, dtype: float64

# 2. String Manipulation
* replace가 아닌 다른 방법을 사용하여 19년도 2분기의 매출액 을 integer로 변경하라.  
결과값은 12578이 나와야함.

In [21]:
df['매출액'][0] = pd.to_numeric(df['매출액'][0])
df['매출액'][0]

12578

# 3. Apply
* 데이터셋의 분기를 제외한 모든 feature에 대해서 수치형 데이터 (int, float) 로 수정하고, describe를 통해 summary statistics 를 확인하라.

* 단 결측치는 0으로 처리한다.

In [22]:
quarter = df['분기']
df = df.drop(['분기'], axis = 1)

In [23]:
def remove_comma(x):
  return x.replace(',', '')

for col in df:
  try: df[col] = pd.to_numeric(df[col].apply(remove_comma))
  except: continue
  
df

매출액  영업이익  영업이익(발표기준)  세전계속사업이익  당기순이익  당기순이익(지배)  당기순이익(비지배)    자산총계  \
0  12578  3991        3991      4452   3245       3228          17  108464   
1  13222  3825        3825      4503   3221       3208          13  106314   
2  11982  2523        2523      1794   1183       1198         -15  107121   
3  11784  3150        3150      4156   2939       2930           9  108594   
4  13188  3947        3947      4020   2935       2931           4  110282   

    부채총계   자본총계  ...  재무활동현금흐름  영업이익률   순이익률  ROE(%)  ROA(%)   부채비율   자본유보율  \
0  25940  82524  ...     -4919  31.73  25.80   12.35    9.37  31.43  803.22   
1  20691  85623  ...      -329  28.93  24.36   12.67    9.95  24.17  836.86   
2  20062  87059  ...      -129  21.06   9.87   12.39    9.96  23.04  850.82   
3  24862  83732  ...       -86  26.73  24.94   13.05    9.79  29.69  823.16   
4  23386  86896  ...     -5501  29.93  22.25   12.20    9.40  26.91  837.40   

   EPS(원)  PER(배)     영업이익률2  
0    2351   13.77  31.730005  
1    2337   13.87  28.929058  
2     873   12.42  21.056585  
3    2134    9.71  26.731161  
4    2135   10.46  29.928723  

[5 rows x 25 columns]

In [24]:
df = df.fillna(0)
df.describe()

매출액         영업이익   영업이익(발표기준)     세전계속사업이익        당기순이익  \
count      5.000000     5.000000     5.000000     5.000000     5.000000   
mean   12550.800000  3487.200000  3487.200000  3785.000000  2704.600000   
std      664.979097   636.931079   636.931079  1131.079131   863.422724   
min    11784.000000  2523.000000  2523.000000  1794.000000  1183.000000   
25%    11982.000000  3150.000000  3150.000000  4020.000000  2935.000000   
50%    12578.000000  3825.000000  3825.000000  4156.000000  2939.000000   
75%    13188.000000  3947.000000  3947.000000  4452.000000  3221.000000   
max    13222.000000  3991.000000  3991.000000  4503.000000  3245.000000   

         당기순이익(지배)  당기순이익(비지배)           자산총계         부채총계         자본총계  ...  \
count     5.000000    5.000000       5.000000      5.00000      5.00000  ...   
mean   2699.000000    5.600000  108155.000000  22988.20000  85166.80000  ...   
std     851.338358   12.481987    1522.396795   2560.38419   1988.82596  ...   
min    1198.000000  -15.000000  106314.000000  20062.00000  82524.00000  ...   
25%    2930.000000    4.000000  107121.000000  20691.00000  83732.00000  ...   
50%    2931.000000    9.000000  108464.000000  23386.00000  85623.00000  ...   
75%    3208.000000   13.000000  108594.000000  24862.00000  86896.00000  ...   
max    3228.000000   17.000000  110282.000000  25940.00000  87059.00000  ...   

         재무활동현금흐름      영업이익률       순이익률     ROE(%)    ROA(%)       부채비율  \
count     5.00000   5.000000   5.000000   5.000000  5.000000   5.000000   
mean  -2192.80000  27.676000  21.444000  12.532000  9.694000  27.048000   
std    2763.51139   4.115651   6.601267   0.335738  0.290224   3.554859   
min   -5501.00000  21.060000   9.870000  12.200000  9.370000  23.040000   
25%   -4919.00000  26.730000  22.250000  12.350000  9.400000  24.170000   
50%    -329.00000  28.930000  24.360000  12.390000  9.790000  26.910000   
75%    -129.00000  29.930000  24.940000  12.670000  9.950000  29.690000   
max     -86.00000  31.730000  25.800000  13.050000  9.960000  31.430000   

           자본유보율       EPS(원)     PER(배)     영업이익률2  
count    5.00000     5.000000   5.000000   5.000000  
mean   830.29200  1966.000000  12.046000  27.675106  
std     18.01957   619.939513   1.898112   4.116711  
min    803.22000   873.000000   9.710000  21.056585  
25%    823.16000  2134.000000  10.460000  26.731161  
50%    836.86000  2135.000000  12.420000  28.929058  
75%    837.40000  2337.000000  13.770000  29.928723  
max    850.82000  2351.000000  13.870000  31.730005  

[8 rows x 25 columns]

In [25]:
df = pd.concat([df, quarter], axis = 1)
df

매출액  영업이익  영업이익(발표기준)  세전계속사업이익  당기순이익  당기순이익(지배)  당기순이익(비지배)    자산총계  \
0  12578  3991        3991      4452   3245       3228          17  108464   
1  13222  3825        3825      4503   3221       3208          13  106314   
2  11982  2523        2523      1794   1183       1198         -15  107121   
3  11784  3150        3150      4156   2939       2930           9  108594   
4  13188  3947        3947      4020   2935       2931           4  110282   

    부채총계   자본총계  ...  영업이익률   순이익률  ROE(%)  ROA(%)   부채비율   자본유보율  EPS(원)  \
0  25940  82524  ...  31.73  25.80   12.35    9.37  31.43  803.22    2351   
1  20691  85623  ...  28.93  24.36   12.67    9.95  24.17  836.86    2337   
2  20062  87059  ...  21.06   9.87   12.39    9.96  23.04  850.82     873   
3  24862  83732  ...  26.73  24.94   13.05    9.79  29.69  823.16    2134   
4  23386  86896  ...  29.93  22.25   12.20    9.40  26.91  837.40    2135   

   PER(배)     영업이익률2     분기  
0   13.77  31.730005   19/6  
1   13.87  28.929058   19/9  
2   12.42  21.056585  19/12  
3    9.71  26.731161   20/3  
4   10.46  29.928723   20/6  

[5 rows x 26 columns]

# 4. NA Value Handling
* 19년도 4분기의 당기순이익(비지배) 부분을 Na로 대체하라.
* 이후 해당 결측치를 mean imputation 방법을 사용하여 처리하라.

In [26]:
df

매출액  영업이익  영업이익(발표기준)  세전계속사업이익  당기순이익  당기순이익(지배)  당기순이익(비지배)    자산총계  \
0  12578  3991        3991      4452   3245       3228          17  108464   
1  13222  3825        3825      4503   3221       3208          13  106314   
2  11982  2523        2523      1794   1183       1198         -15  107121   
3  11784  3150        3150      4156   2939       2930           9  108594   
4  13188  3947        3947      4020   2935       2931           4  110282   

    부채총계   자본총계  ...  영업이익률   순이익률  ROE(%)  ROA(%)   부채비율   자본유보율  EPS(원)  \
0  25940  82524  ...  31.73  25.80   12.35    9.37  31.43  803.22    2351   
1  20691  85623  ...  28.93  24.36   12.67    9.95  24.17  836.86    2337   
2  20062  87059  ...  21.06   9.87   12.39    9.96  23.04  850.82     873   
3  24862  83732  ...  26.73  24.94   13.05    9.79  29.69  823.16    2134   
4  23386  86896  ...  29.93  22.25   12.20    9.40  26.91  837.40    2135   

   PER(배)     영업이익률2     분기  
0   13.77  31.730005   19/6  
1   13.87  28.929058   19/9  
2   12.42  21.056585  19/12  
3    9.71  26.731161   20/3  
4   10.46  29.928723   20/6  

[5 rows x 26 columns]

In [27]:
df['당기순이익(비지배)'][2] = pd.NA
df

매출액  영업이익  영업이익(발표기준)  세전계속사업이익  당기순이익  당기순이익(지배) 당기순이익(비지배)    자산총계  \
0  12578  3991        3991      4452   3245       3228         17  108464   
1  13222  3825        3825      4503   3221       3208         13  106314   
2  11982  2523        2523      1794   1183       1198       <NA>  107121   
3  11784  3150        3150      4156   2939       2930          9  108594   
4  13188  3947        3947      4020   2935       2931          4  110282   

    부채총계   자본총계  ...  영업이익률   순이익률  ROE(%)  ROA(%)   부채비율   자본유보율  EPS(원)  \
0  25940  82524  ...  31.73  25.80   12.35    9.37  31.43  803.22    2351   
1  20691  85623  ...  28.93  24.36   12.67    9.95  24.17  836.86    2337   
2  20062  87059  ...  21.06   9.87   12.39    9.96  23.04  850.82     873   
3  24862  83732  ...  26.73  24.94   13.05    9.79  29.69  823.16    2134   
4  23386  86896  ...  29.93  22.25   12.20    9.40  26.91  837.40    2135   

   PER(배)     영업이익률2     분기  
0   13.77  31.730005   19/6  
1   13.87  28.929058   19/9  
2   12.42  21.056585  19/12  
3    9.71  26.731161   20/3  
4   10.46  29.928723   20/6  

[5 rows x 26 columns]

In [28]:
df['당기순이익(비지배)'][2] = df['당기순이익(비지배)'].mean()
df

매출액  영업이익  영업이익(발표기준)  세전계속사업이익  당기순이익  당기순이익(지배) 당기순이익(비지배)    자산총계  \
0  12578  3991        3991      4452   3245       3228         17  108464   
1  13222  3825        3825      4503   3221       3208         13  106314   
2  11982  2523        2523      1794   1183       1198      10.75  107121   
3  11784  3150        3150      4156   2939       2930          9  108594   
4  13188  3947        3947      4020   2935       2931          4  110282   

    부채총계   자본총계  ...  영업이익률   순이익률  ROE(%)  ROA(%)   부채비율   자본유보율  EPS(원)  \
0  25940  82524  ...  31.73  25.80   12.35    9.37  31.43  803.22    2351   
1  20691  85623  ...  28.93  24.36   12.67    9.95  24.17  836.86    2337   
2  20062  87059  ...  21.06   9.87   12.39    9.96  23.04  850.82     873   
3  24862  83732  ...  26.73  24.94   13.05    9.79  29.69  823.16    2134   
4  23386  86896  ...  29.93  22.25   12.20    9.40  26.91  837.40    2135   

   PER(배)     영업이익률2     분기  
0   13.77  31.730005   19/6  
1   13.87  28.929058   19/9  
2   12.42  21.056585  19/12  
3    9.71  26.731161   20/3  
4   10.46  29.928723   20/6  

[5 rows x 26 columns]

## Feature Engineering
Relative Perfomance 라는 새로운 feature를 계산하라.

이는 최근 1년치 매출액의 평균값을 기준으로

* 10% 이상 -> S
* 5% 이상 -> A
* -5 ~ 5% -> B
* -5%이하 -> C
* 10%이하 -> D
라는 값을 갖는 feature이다.

20년도 2분기에 해당하는 결과값은 A가 나와야함.

각각에 해당하는 등급이 나오기 위해서 필요한 매출액을 추가로 서술하라.

In [29]:
M = sum(df['매출액'][1:5]) / 4
M

12544.0

In [30]:
t = df['매출액'][4]
t

13188

In [31]:
# 20년도 2분기 RP 값 구하는 식

RP = ((t-M)/M ) * 100
RP

"""
역산하여 각 등급을 받기 위한 매출액을 구하기 위한 식은

t(해당 분기 매출액) = (RP / 100) * M + M

M은 1년치 평균 매출액

따라서

S 등급을 받기 위해 더 필요한 매출액
"""

M = sum(df['매출액'][1:5]) / 4
t_S = (10 / 100) * M + M
RP_S = t_S - t

RP_S

610.3999999999996

In [32]:
# B 등급 기준 초과한 매출액

t_B_max = (5 / 100) * M + M
t_B_min = (-5 / 100) * M + M
RP_B_max = t_B_max - t
RP_B_min = t_B_min - t

print(RP_B_max, RP_B_min)

-16.799999999999272 -1271.2000000000007


In [33]:
# C 등급 기준 초과한 매출액

t_C = (-5 / 100) * M + M
RP_C = t_C - t

RP_C

-1271.2000000000007

In [34]:
# D 등급 기준 초과한 매출액

t_D = (-10 / 100) * M + M
RP_D = t_D - t

RP_D

-1898.3999999999996